Implement Python Array API `asarray` function.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #60627

In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

Differential Revision: [D31640510](https://our.internmc.facebook.com/intern/diff/D31640510)

ysiraichi committed:
Implement Python Array API `asarray` function.

In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+asarray(obj, *, dtype=None, device=None, copy=None, requires_grad=False) -> Tensor
+
+Creates a new :class:`Tensor` out of :attr:`obj` storage, by either copying or
+aliasing it.
```
nit -- it might be clearer to use "view" instead of "alias"

mruberry replied:
To be consistent with the `as_tensor` documentation (which is being updated):


"Converts :attr:`obj` into a tensor, sharing data and preserving the autograd graph if possible."

Follow-up question: is this an accurate description? See inline comment above about autograd graph preservation.

ysiraichi replied:
As is, `asarray` preserves the autograd graph only if we create a view of the input tensor. I'm explicitly detaching the copied tensors. That said, I agree it's better to be consistent with `as_tensor`.

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+3. a Python object that implements the buffer protocol (copy and alias allowed)
+4. an arbitrary sequence (only copy allowed)
+
+When :attr:`copy` is not given, this function will try first to alias the given
```
same as above

mruberry replied:
I agree with @anjali411 that "view" is more familiar than the term "alias" here. But I also think this detail can be moved into the arg's description below

anjali411 commented:
For the cases where `as_array` performs a view, we should error out if the input has conjugate or negative bit set. see example PR fixing this for `.numpy()`: https://github.com/pytorch/pytorch/pull/61925

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi commented:
@anjali411 Thank you for your observations. I hadn't really thought about views, here. :)

> For the cases where as_array performs a view, we should error out if the input has conjugate or negative bit set. see example PR fixing this for .numpy(): #61925

I'm not sure this applies for `asarray`.
As I understand it, we should error out if we are creating a view out of `obj` with some different dtype, where `obj` is a tensor, right? If that's the case, it would only happen if we called `torch.asarray(tensor, dtype=<different-dtype>)`. But, in this case, we never create a view. We either:

1. copy `tensor` if `copy=None` or `copy=True`, interpreting its values as `<different-dtype>`
2. we error out if `copy=False` (we don't even try to create a view here)

Regarding (2): not sure if we should, though. I think it makes more sense to leave it to the specialized functions (e.g. `torch.Tensor.view`), while `asarray` focus only on creating tensors out of other stuff.
@mruberry @rgommers What are your thoughts?

ysiraichi commented:
@mruberry @rgommers This is a friendly reminder. Do you have some time to take a look at this PR?

mruberry commented:
> @mruberry @rgommers This is a friendly reminder. Do you have some time to take a look at this PR?

Hey @ysiraichi! Thanks for the ping. In the future if I don't get back to you within 2 business days please ping me again. Some notifications get lost. I'll look at the PR now.

mruberry commented on test/run_test.py:
```diff
@@ -40,9 +40,9 @@
     'test_type_hints',
     'test_ao_sparsity',
     'test_autograd',
+    'test_asarray',
```
Instead of creating a new test file let's add this to test_tensor_creation_ops

mruberry replied:
(we can move the buffer tests there, too, and kill that separate test file)

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        return torch.from_numpy(o)
+    raise RuntimeError(f"object {o} not supported.")
+
+# Tests for the `asarray` function:
```
Nice comment

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+#   The implementation itself is based on the Python Array API:
+#   https://data-apis.org/array-api/latest/API_specification/creation_functions.html
+class TestAsArray(common.TestCase):
+    def _check(self, original, cvt=lambda t: t, is_alias=True, same_dtype=True, same_device=True, **kwargs):
```
Nicely written check function

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+    def test_alias_from_tensor(self, device, dtype):
+        # DLpack does not explicitly support bool
+        # It does it through uint8 type
+        if dtype is torch.bool:
```
Nit: skip the test instead of just returning (here and in the other tests)

mruberry replied:
Also: while this is true, DLPack isn't tested here, is it? So maybe this skip is not needed

ysiraichi replied:
Ah, you are right. I had completely missed that.

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        def check(**kwargs):
+            self._check(original, is_alias=False, **kwargs)
+
+        check(requires_grad=self._may_compute_grad(dtype), copy=True)
```
Check the requires_grad -> requires_grad and requires_grad -> no_grad case, too

mruberry replied:
Also - is the computational graph retained when this function is called? That is, will grad propagate from the tensor created using asarray, whether it's an "alias" or not, back to the original tensor if the original also requires_grad?

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+    @skipMeta
+    @dtypes(*get_all_dtypes())
+    def test_copy_tensor(self, device, dtype):
+        if dtype is torch.bool:
```
I think bool can be tested here, too

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+
+        # Copy is forced because of different device
+        if torch.cuda.is_available():
+            other = "cuda" if device == "cpu" else "cpu"
```
Assert an error here if `copy=False`

mruberry replied:
Don't compare device to a string, instead use `self.device_type` or `torch.device(device).type`

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        check(device=device, dtype=dtype, copy=True)
+
+        # Copy is forced because of different device
+        if torch.cuda.is_available():
```
Add a cuda:0 -> cuda:1 version of this test, too, by creating a test with the decorator

https://github.com/pytorch/pytorch/blob/22f78144c70d9d215bc390805aa8d0657e548b85/torch/testing/_internal/common_device_type.py#L922

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        for other in get_all_dtypes():
+            if dtype != other:
+                check(same_dtype=False, dtype=other)
+                check(same_dtype=False, dtype=other, copy=True)
```
Assert what happens if copy=False here, too

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+    def test_alias_from_dlpack(self, device, dtype):
+        # DLpack does not explicitly support bool
+        # It does it through uint8 type
+        if dtype is torch.bool:
```
This skip for bool is needed

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+
+    @skipMeta
+    @dtypes(*get_all_dtypes())
+    def test_copy_from_dlpack(self, device, dtype):
```
The design of these tests is very readable and makes a lot of logical sense

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+
+    @dtypes(*get_all_dtypes())
+    def test_copy_list(self, device, dtype):
+        original = common.make_tensor((5, 5), device, dtype)
```
Might as well always create this tensor on the cpu since it's going to be converted to a list

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        original = common.make_tensor((5, 5), device, dtype)
+
+        if torch.cuda.is_available():
+            other_device = "cuda" if device == "cpu" else "cpu"
```
See previous comment on device type checking

mruberry commented on test/test_asarray.py:
```diff
@@ -0,0 +1,423 @@
+        if torch.cuda.is_available():
+            other_device = "cuda" if device == "cpu" else "cpu"
+
+            with self.assertRaisesRegex(ValueError,
```
Previously I suggested asserts like this appear earlier, but it's OK to collect them here, too

mruberry commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -485,73 +485,37 @@ static PyObject * THPVariable_frombuffer(PyObject* self_, PyObject* args, PyObje
-        count > 0 || (len - offset) % elsize == 0,
-        "buffer length (", len - offset, " bytes) after offset (", offset, " bytes) "
-        "must be a multiple of element size (", elsize, ")");
+static PyObject * THPVariable_asarray(PyObject* self_, PyObject* args, PyObject* kwargs)
```
Why does `asarray` need custom parsing?

ysiraichi replied:
By "custom parsing", do you mean the fact that I'm using `PythonArgParser` here? If so, it's not really related to `self`. It's just a way I found to parse the `kwargs`.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+Creates a new :class:`Tensor` out of :attr:`obj` storage, by either copying or
+aliasing it.
+
+:attr:`obj` will be tested for, in order:
```
":attr:\`obj\` can be one of:"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+
+:attr:`obj` will be tested for, in order:
+
+1. a :class:`Tensor` object (copy and alias allowed)
```
"a tensor"

mruberry replied:
Can this convert NumPy arrays?

ysiraichi replied:
Yes, but it handles NumPy arrays as python sequences. In other words, it always copies from NumPy arrays. Should we allow their storage to be shared (it should be straight-forward, since we already have `from_numpy`)?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+:attr:`obj` will be tested for, in order:
+
+1. a :class:`Tensor` object (copy and alias allowed)
+2. a DLPack object (copy and alias allowed)
```
"a DLPack capsule"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+
+1. a :class:`Tensor` object (copy and alias allowed)
+2. a DLPack object (copy and alias allowed)
+3. a Python object that implements the buffer protocol (copy and alias allowed)
```
Just remove the parenthetical expression here

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+1. a :class:`Tensor` object (copy and alias allowed)
+2. a DLPack object (copy and alias allowed)
+3. a Python object that implements the buffer protocol (copy and alias allowed)
+4. an arbitrary sequence (only copy allowed)
```
"a Python Sequence"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+2. a DLPack object (copy and alias allowed)
+3. a Python object that implements the buffer protocol (copy and alias allowed)
+4. an arbitrary sequence (only copy allowed)
+
```
This is a good time to explain a few things:

- how does dtype work with DLPack capsules and Python buffer protocol objects?
- how does requires_grad work? See https://github.com/pytorch/pytorch/pull/63308/files, which is updating tensor and as_tensor to clarify their explanations, too

ysiraichi replied:
I will add an explanation for those things.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+vice-versa.
+
+.. warning::
+    When passing a Python object that implements the buffer protocol, this function
```
Good warning. Should this also have a pointer to frombuffer?

ysiraichi replied:
I have referenced `frombuffer` in the `seealso`section. Do you think I should do it when talking about how it uses the provided `dtype`, too?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+    When passing a Python object that implements the buffer protocol, this function
+    will not infer its type, but will assume that the type is the given :attr:`dtype`.
+
+.. warning::
```
I don't think this warning is necessary

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+    - :attr:`copy` is ``False``, but the given :attr:`obj` is neither a
+      :class:`Tensor`, a DLPack object, nor a Python object that implements the
+      buffer protocol.
+
```
Add a seealso section like svd has (https://pytorch.org/docs/master/generated/torch.linalg.svd.html?highlight=svd#torch.linalg.svd) for `torch.tensor` and `torch.as_tensor` -- how is this different from them?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+
+Keyword args:
+    {dtype}
+    copy (bool, optional): defines whether to copy, alias, or whatever is
```
"If None then the result tensor shares its storage with the Python object when possible. If True then storage is never shared. If False storage is always shared. If the storage cannot be shared and this is False then an error is thrown."

Is the case where `copy=False` but no copy can be made tested?

ysiraichi replied:
> Is the case where copy=False but no copy can be made tested?

Yes! All the cases where `copy=False`, even though we need to do it, are tested on `test_unsupported_alias`.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+    {dtype}
+    copy (bool, optional): defines whether to copy, alias, or whatever is
+        appropriate (aliasing is prioritized).
+    {device}
```
The default device text isn't appropriate here since the device is inferred from the object. See the updated tensor and as_tensor docs (PR linked above) for an example

mruberry commented on torch/_torch_docs.py:
```diff
@@ -970,6 +970,64 @@ def merge_dicts(*dicts):
+
+Example::
+
+    >>> a = torch.randint(10, (10, 10), dtype=torch.int32)
```
Add comments for these examples

mruberry reviewed (commented):
Hey @ysiraichi!

I apologize again for the delay. This is a very detail PR and I appreciate your thoroughness. I made a lot of detailed inline comments, too, and I look forward to hearing your thoughts on them. Please let me know if you have any questions. I will respond to you quickly going forward.

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi commented:
Hi @mruberry, thank you for your review (and sorry for the delay).
I have addressed your comments and changed the PR accordingly. I will summarize the changes below:

- Improved the documentation for better consistency with `as_tensor`, `from_dlpack`, and `from_numpy`
- Moved tests to `test_tensor_creation_ops.py`, and deleted `test_asarray.py`
- Added support for sharing memory with NumPy
- Improved testing code
    - Unified tests that are testing the same thing (either sharing or copying memory)
    - Fixed (correct and incorrect) dtype skipping

Again, sorry for the delay. Let me know what you think (and whether there's anything missing)!

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

[ghstack-poisoned]

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -948,6 +948,101 @@ def merge_dicts(*dicts):
+           r"""
+asarray(obj, *, dtype=None, device=None, copy=None, requires_grad=False) -> Tensor
+
+Converts :attr:`obj` into a tensor, sharing data and preserving autograd history
```
could you please disable `asarray` for tensors with conj and neg bit set to `True`? see https://github.com/pytorch/pytorch/pull/61925 which disables `numpy()` for similar cases

ysiraichi replied:
Hi @anjali411, thank you for the review.

Maybe I'm missing something here, but I don't think we need it here.
As far as I understand it, it should be disabled if we are: (1) transforming the tensor into something else (e.g. NumPy array); or (2) interpreting the tensor bytes as some other dtype (e.g. by calling `view(dtype)`). Is that correct?

If so, this function does neither. If the input is a tensor, we wrap the same tensor into another Python object (i.e. no view with different dtypes are created). If a different dtype is provided, it calls `tensor.to(other_dtype)` (which, I believe, does not disables those tensors).

I'm not too familiar with those bit flags, so the problem might be that I just don't understand how they work. Sorry about that.
Could you explain to me a little bit more the reason why this function needs to be disabled for those tensors?

anjali411 replied:
@ysiraichi `tensor.numpy()` creates a view, meaning it returns a numpy array object that's viewing the same memory as `tensor`. I assume this to be true for `asarray` as well. A simple way to test this would be to perform an in-place operation on the `tensor` after `asarray` is called:

```py
x = torch.randn(2)
y = x.asarray()
x.zero_()
# if it prints zero, then it validates that y views the same memory as x
print(y) 
```

anjali411 reviewed (changes requested):
Please disable `asarray` for tensors with conj and neg bit set as indicated in the comment

ysiraichi commented:
Hi, @anjali411 @mruberry. This is a friendly reminder about this PR. Do you have some time to take a look at it?

mruberry commented:
> Hi, @anjali411 @mruberry. This is a friendly reminder about this PR. Do you have some time to take a look at it?

Of course! Thank you for the ping, @ysiraichi. We have a holiday weekend now but I'll make time to look at this ASAP!

ysiraichi commented:
Hi, @anjali411 @mruberry. This is a friendly reminder about this PR. Do you have some time to take a look at it?

ysiraichi commented:
Hi, @anjali411 @mruberry. This is a friendly reminder about this PR. Do you have some time to take a look at it?

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -948,6 +948,101 @@ def merge_dicts(*dicts):
+    >>> a = torch.tensor([1, 2, 3])
+    >>> # Shares memory with tensor 'a'
+    >>> b = torch.asarray(a)
+    >>> a.data_ptr() == b.data_ptr()
```
okay yes so based on this, we should disable `asarray` for conjugated and negated tensors (see the comment above pointing to a PR that made a similar change for `.numpy()`)

anjali411 commented on test/test_tensor_creation_ops.py:
```diff
@@ -3425,9 +3428,416 @@ def test_full_like_inference(self, device):
+def get_dtype_size(dtype):
+    return int(torch.empty((), dtype=dtype).element_size())
+
+class TestBufferProtocol(TestCase):
```
was this just verbatim moved from the file deleted above?

anjali411 reviewed (approved):
LGTM overall! It might be nice to add an `OpInfo` for `asarray`

cc. @mruberry for a final review

ysiraichi commented:
Hi, @mruberry. This is a friendly reminder about this PR. Do you have some time to take a look at it?

mruberry reviewed (approved):
Hey @ysiraichi!

Thank you for your patience and pings. I finally had a chance to read through this PR again thoroughly (I tried several times but something always came up!). It's very good, and I appreciate @anjali411's concern but don't think this PR suffers from the issue. There are a few small tweaks I would like to suggest about readability and docs, but they're minor and I think we should land this as-is. Nice work, this is a challenging, detailed PR that required a great deal of effort, thoughtfulness, and attention to detail.

mruberry commented:
Ack! Looks like this just needs a rebase, @ysiraichi. Sorry about that. Just ping me once it's rebased.

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

Differential Revision: [D31640510](https://our.internmc.facebook.com/intern/diff/D31640510)

[ghstack-poisoned]

ysiraichi commented:
@mruberry Rebased!

ysiraichi committed:
Update on "Implement Python Array API `asarray` function."


In this PR, the core of `frombuffer` and `fromDLPack` onto _tensor_new.cpp_. `asarray`
uses such refactored functions for interpreting the object as a tensor. We follow the
Python Array API standard found:

https://data-apis.org/array-api/latest/API_specification/creation_functions.html?highlight=asarray

Differential Revision: [D31640510](https://our.internmc.facebook.com/intern/diff/D31640510)

[ghstack-poisoned]
